File: neetcode-cpp/BalancedBinaryTree.cpp
#include <iostream>
#include <vector>

#include "TreeNode.h"
using namespace std;

class BalancedBinaryTree {
private:
    vector<int> dfs(const TreeNode *root) {
        if (root == nullptr) {
            return {1, 0};
        }

        vector<int> left = dfs(root->left);
        vector<int> right = dfs(root->right);
        bool isBalanced = left[0] == 1 && right[0] == 1 && abs(left[1] - right[1]) <= 1;
        int height = 1 + max(left[1], right[1]);

        return {isBalanced ? 1 : 0, height};
    }

public:
    bool isBalanced(TreeNode *root) { return dfs(root)[0] == 1; }
};

int main() {
    auto tree = TreeNode(1);
    tree.left = new TreeNode(2);
    tree.right = new TreeNode(3);
    tree.left->left = new TreeNode(4);
    tree.left->right = new TreeNode(5);
    tree.right->left = new TreeNode(6);
    tree.right->right = new TreeNode(7);

    cout << BalancedBinaryTree().isBalanced(&tree) << endl;

    const auto tree2 = new TreeNode(1);
    tree2->left = new TreeNode(2);
    tree2->left->left = new TreeNode(3);
    tree2->left->right = new TreeNode(3);
    tree2->left->left->left = new TreeNode(4);
    tree2->left->left->right = new TreeNode(4);
    tree2->left->left->left->left = new TreeNode(5);

    cout << BalancedBinaryTree().isBalanced(tree2) << endl;

    return 0;
}

File: neetcode-kotlin/app/src/main/kotlin/com/niyaj/neetcode/BalancedBinaryTree.kt
package com.niyaj.neetcode

import kotlin.math.abs

/**
 * Balanced Binary Tree
 * Given a binary tree, return true if it is height-balanced and false otherwise.
 *
 * A height-balanced binary tree is defined as a binary tree in which the left and right subtrees of every node differ in height by no more than 1.
 *
 * Example 1:
 *
 *
 *
 * Input: root = [1,2,3,null,null,4]
 *
 * Output: true
 * Example 2:
 *
 *
 *
 * Input: root = [1,2,3,null,null,4,null,5]
 *
 * Output: false
 * Example 3:
 *
 * Input: root = []
 *
 * Output: true
 * Constraints:
 *
 * The number of nodes in the tree is in the range [0, 1000].
 * -1000 <= Node.val <= 1000
 */
object BalancedBinaryTree {
    fun isBalanced(root: TreeNode<Int?>?): Boolean {
        return dfs(root).first
    }

    /**
     * Depth First Search
     * Step1: If root is null, return true and 0
     * Step2: Recursively call dfs for left and right nodes
     * Step3: Check if left and right nodes are balanced and the difference between their heights is less than or equal to 1
     * Step4: Return the result and the height of the current node
     * Step5: Repeat the above steps for all the nodes
     * Step6: Return the result
     *
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     *
     *
     * @param root TreeNode
     * @return Pair<Boolean, Int>
     *
     */
    private fun dfs(root: TreeNode<Int?>?): Pair<Boolean, Int> {
        if (root == null) {
            return Pair(true, 0)
        }

        val left = dfs(root.left)
        val right = dfs(root.right)

        val isBalanced = left.first && right.first && abs(left.second - right.second) <= 1
        val height = maxOf(left.second, right.second) + 1
        return Pair(isBalanced, height)
    }
}

fun main() {
    val root: TreeNode<Int?> = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left?.left = null
    root.left?.right = null
    root.right?.left = TreeNode(4)
    root.right?.right = null
    root.right?.left?.left = TreeNode(5)

    println(BalancedBinaryTree.isBalanced(root))

    val root2: TreeNode<Int?> = TreeNode(null)
    println(BalancedBinaryTree.isBalanced(root2))

}